src/sidebar/build-thread.js

```diff
@@ -2,107 +2,106 @@
+ *       children. This count includes visually-hidden threads.
+ * @prop {number} replyCount - Computed count of all replies to a thread
+ * @prop {'dim'|'highlight'} [highlightState] - In cases where there are one
+ *       or more threads currently designated as "highlighted", we track
+ *       the highlight state for each thread. undefined when there are no
+ *       highlighted threads.
  */
 
 /**
- * Default state for new threads, before applying filters etc.
- *
- * @type {Thread}
+ * Default state for new threads
  */
 const DEFAULT_THREAD_STATE = {
-  /**
-   * The ID of this thread. This will be the same as the annotation ID for
-   * created annotations or the `$tag` property for new annotations.
-   */
   id: '__default__',
-  /**
-   * The Annotation which is displayed by this thread.
-   *
-   * This may be null if the existence of an annotation is implied by the
-   * `references` field in an annotation but the referenced parent annotation
-   * does not exist.
-   */
-  annotation: undefined,
-  /** The parent thread ID */
-  parent: undefined,
-  /** True if this thread is collapsed, hiding replies to this annotation. */
   collapsed: false,
-  /** True if this annotation matches the current filters. */
   visible: true,
-  /** Replies to this annotation. */
-  children: [],
-  /**
-   * The total number of children of this annotation,
-   * including any which have been hidden by filters.
-   */
+  replyCount: 0,
   totalChildren: 0,
-  /**
-   * The highlight state of this annotation:
-   *  undefined - Do not (de-)emphasize this annotation
-   *  'dim' - De-emphasize this annotation
-   *  'highlight' - Emphasize this annotation
-   */
-  highlightState: undefined,
 };
 
 /**
  * Returns a persistent identifier for an Annotation.
  * If the Annotation has been created on the server, it will have
- * an ID assigned, otherwise we fall back to the local-only '$tag'
+ * an id assigned, otherwise we fall back to the local-only '$tag'
  * property.
+ *
+ * @param {Annotation} annotation
+ * @return {string}
  */
-function id(annotation) {
+function annotationId(annotation) {
   return annotation.id || annotation.$tag;
 }
 
 /**
- * Link the annotation with ID `id` to its parent thread.
+ * Is there a valid path from the thread indicated by `id` to the root thread,
+ * with no circular references?
  *
+ * @param {string} id - The id of the thread to be verified
+ * @param {string} ancestorId - The ancestor of the thread indicated by id that
+ *        is to be verified: is it extant and not a circular reference?
+ * @return {boolean}
+ */
+function hasPathToRoot(threads, id, ancestorId) {
+  if (!threads[ancestorId] || threads[ancestorId].parent === id) {
+    // Thread for ancestor not found, or points at itself: circular reference
+    return false;
+  } else if (!threads[ancestorId].parent) {
+    // Top of the tree: we've made it
+    return true;
+  }
+  return hasPathToRoot(threads, id, threads[ancestorId].parent);
+}
```

Comment:
In thinking through this I found a test case which results in an infinite loop on `master` and infinite recursion in `hasPathToRoot` on this branch:

```js
    it('handles annotations that have their own ID in their `references` list', () => {
      const fixture = [
        { id: '1', references: ['1'] },
        { id: '2', references: ['1'] },
      ];
      const thread = createThread(fixture);
      assert.deepEqual(thread, []);
    });
```

Reply:
If the client ever encountered the annotations from this fixture that would indicate that something has gone elsewhere in the Hypothesis system, but I don't know if anything actually prevents such annotations from being created on the server.

Reply:
I have a fix for this, but just FYI that I've updated the test assert to:

```
      assert.deepEqual(thread, [
        {
          annotation: {
            id: '1',
            references: ['1'],
          },
          children: [
            {
              annotation: {
                id: '2',
                references: ['1'],
              },
              children: [],
            },
          ],
        },
      ]);
```